### ISSUE TITLE
_pre_test decorator alters individual multiple times

### ISSUE THREAD

**Issue by mficek** [rhiever/tpot#368]
`_pre_test` decorator calls passed function and tests, if the resulting pipeline is a resonable (working) one. However, if not, the next function call uses a mutated/mated individual and not the original one.

## Context of the issue

When mutation happens via `ind_out = toolbox.mutate(ind)`, the `ind_out` is identical with `ind` when function ends. Therefore we have to call `ind = toolbox.clone(ind_original)` if mutate runs in inside a cycle.

This clonning behaviour is missing in `_pre_test` decorator.  

## Process to reproduce the issue

1. See all necessary changes (debug printounts) in 
[git diff](https://github.com/rhiever/tpot/compare/development...mficek:mut_check)

2. Run with verbosity=3 on dataset with less then 200 features. To explain: I use only one classifier and set one of it's parameters to [10, 200] in argument list. That leads to errors during pipeline validation, thus aplying funciton inside `_pre_test` more often. 

3. Look for debug output where mutation applied to a pipeline does not correspond with the output. (see more bellow)

## Current result
```
2 ==+> Mutation DecisionTreeClassifier(input_matrix, 'entropy', 7, 10, 14, 15)
mut: replacement
_pre_test decorator: _random_mutation_operator: num_test=0 max_features must be in (0, n_features]
_pre_test expr: DecisionTreeClassifier(input_matrix, 'gini', 9, 200, 6, 17)
mut: insert
_pre_test decorator: _random_mutation_operator: num_test=1 max_features must be in (0, n_features]
_pre_test expr: DecisionTreeClassifier(input_matrix, 'gini', 9, 200, 6, 17)
mut: shrink
_pre_test decorator: _random_mutation_operator: num_test=2 max_features must be in (0, n_features]
_pre_test expr: DecisionTreeClassifier(input_matrix, 'gini', 9, 200, 6, 17)
mut: insert
_pre_test decorator: _random_mutation_operator: num_test=3 max_features must be in (0, n_features]
_pre_test expr: DecisionTreeClassifier(input_matrix, 'gini', 9, 200, 6, 17)
mut: replacement
_pre_test decorator: _random_mutation_operator: num_test=4 max_features must be in (0, n_features]
_pre_test expr: DecisionTreeClassifier(input_matrix, 'gini', 8, 200, 6, 17)
mut: insert
_pre_test decorator: _random_mutation_operator: num_test=5 max_features must be in (0, n_features]
_pre_test expr: DecisionTreeClassifier(input_matrix, 'gini', 8, 200, 6, 17)
mut: replacement
_pre_test decorator: _random_mutation_operator: num_test=6 max_features must be in (0, n_features]
_pre_test expr: DecisionTreeClassifier(input_matrix, 'entropy', 8, 200, 6, 17)
mut: replacement
_pre_test decorator: _random_mutation_operator: num_test=7 max_features must be in (0, n_features]
_pre_test expr: DecisionTreeClassifier(input_matrix, 'entropy', 9, 200, 6, 17)
mut: replacement
2 ==-> Mutation DecisionTreeClassifier(input_matrix, 'entropy', 9, 10, 6, 17)
```
You can follow the workflow: an Individual enters a mutation loop and at some point the _pre_test fails, founds an invalid pipeline. But the inner cycle of _pre_test uses mutated individual as input.  

## Possible fix

`_pre_test` decorates `_generate`, mutate and mate. Since _generate creates a new Individual, `_pre_test` works. I suggest to create new decorator that would handle exclusively mate and mutate in such a way that 

1. the function arguments are always individuals
2. all individuals are clonned before function is called

Something of this kind:
```
---> def _op_pre_test(func):  # different name
    @wraps(func)
    def check_pipeline(self, *args, **kwargs):
        bad_pipeline = True
        num_test = 0 # number of tests
        while bad_pipeline and num_test < 10: # a pool for workable pipeline
--->            individuals = [toolbox.clone(ind) for ind in args]
            try:
                with warnings.catch_warnings():
                    warnings.simplefilter('ignore')
                    expr = func(self, individuals, **kwargs)
--->                    for expr_test in expr:  # here the expr will be ALWAYS a tuple
                        #print(num_test, generate_pipeline_code(expr_to_tree(expr), self.operators)) # debug
                        sklearn_pipeline = eval(generate_pipeline_code(expr_to_tree(expr_test), self.operators), self.operators_context)
                        if self.classification:
                            sklearn_pipeline.fit(pretest_X, pretest_y)
                        else:
                            sklearn_pipeline.fit(pretest_X_reg, pretest_y_reg)
                        bad_pipeline = False
            except BaseException as e:
                if self.verbosity == 3:
                    print('_pre_test decorator: {fname}: num_test={n} {e}'.format(n=num_test, fname=func.__name__, e=e))
                    print('_pre_test expr: {expr}'.format(expr = expr[0]))
                pass
            finally:
                num_test += 1
        return expr
    return check_pipeline
```
What do you think? Can you see some elegant way, how to modify `_pre_test` behavior based on function decorated?

**Comment by weixuanfu2016** [comment]
Thank you for this good idea. I think, for adding `toolbox.clone` , we could edit `_pre_test` in your PR #367 a little with `isinstance(ind, creator.Individual)` as below:

```
from deap import creator

def check_pipeline(self, *args, **kwargs):
    bad_pipeline = True
    num_test = 0 # number of tests
    while bad_pipeline and num_test < 10: # a pool for workable pipeline
--->        args = [self._toolbox.clone(arg) if isinstance(arg, creator.Individual) else arg for arg in args]
        try:
            with warnings.catch_warnings():
                warnings.simplefilter('ignore')
                expr = func(self, *args, **kwargs)


```

**Comment by mficek** [comment]
I agree, it looks good. I added this into #367. Thanks!

**Comment by mficek** [comment]
Resolved in #367 and merged. Thanks! Closing.
